ext/ddtrace_profiling_native_extension/private_vm_api_access.c

```diff
@@ -49,6 +50,92 @@ rb_nativethread_id_t pthread_id_for(VALUE thread) {
   #endif
 }
 
+// Queries if the current thread is the owner of the global VM lock.
+//
+// @ivoanjo: Ruby has a similarly-named `ruby_thread_has_gvl_p` but that API is insufficient for our needs because it can
+// still return `true` even when a thread DOES NOT HAVE the global VM lock.
+// In particular, looking at the implementation, that API assumes that if a thread is not in a "blocking region" then it
+// will have the GVL which is probably true for the situations that API was designed to be called from BUT this assumption
+// does not hold true when calling `ruby_thread_has_gvl_p` from a signal handler. (Because the thread may have lost the
+// GVL due to a scheduler decision, not because it decided to block.)
+//
+// Thus we need our own gvl-checking method which actually looks at the gvl structure to determine if it is the owner.
+bool is_current_thread_holding_the_gvl(void) {
+  current_gvl_owner owner = gvl_owner();
+  return owner.valid && pthread_equal(pthread_self(), owner.owner);
+}
+
+#ifndef NO_GVL_OWNER // Ruby < 2.6 doesn't have the owner/running field
+  current_gvl_owner gvl_owner(void) {
+    // TODO: Reading owner/running is a racy read. Should we perhaps add a read barrier here?
```

Comment:
Will a barrier help here, given the data source (owner/running) won't have any synchronization with `is_current_thread_holding_the_gvl`?

Reply:
It might ok to just say something like "we can't synchronize with the data source, so this read is racy"

Reply:
You're right; this is always racy so unless we do decide to acquire the lock that Ruby uses to protect the `owner`, it will always be possible to have a racy read.

To be honest, the racy read is actually not a problem in this case:
* Only a thread sets and unset itself as the owner, so if a thread sees itself as the owner, that will be correct
* In a future PR I'll use `gvl_owner` to target the SIGPROF signals in the profiler, but in that case sending a signal to another thread is always a race so if we read something outdated here it's no problem, that's why the signal handler checks if it landed on the right thread.

I'll update the comment to reflect this discussion 👍

Reply:
Comment added in d9b420c3e6024b9fa6a7a6f7c4acc6197dc901fc